### PR TITLE
fix(parser) Parse tags_key/tags_value as FunctionCall and not Column

### PR DESCRIPTION
**Pull request by fpacifici**
It seems to me that, as long as there won't be any much deeper refactoring of tags_key/value fields that would change the query language as well, it would be more appropriate to parse `tag_keys` and `tag_values` as FunctionCall and not Columns since they are not Columns, they do not produce one result per record but basically a join expression that is then transformed into a different function call.

### CODE REVIEW

**Comment by fpacifici** [comment]
Also, we may consider making this a keyword (it kind of is in the snuba language) still `tags_key` and `tags_value` are not supported in all environments (as a keyword should be). They are dataset/storage specific since there has to be a tags column. So in today's data model it seems more like a function with a special syntax in the language. We already have dataset specific functions like impact and apdex.

**Comment by lynnagara** [comment]
Well I agree they are not columns, but FunctionCall implies to me a call to a built in ClickHouse function which these are also not. Is it worth introducing a separate type of expression for these instead?

**Comment by fpacifici** [comment]
"FunctionCall implies to me a call to a built in ClickHouse function which these are also not"
I don't think this is meant to be the case. the parser generates the LogicalQuery, we can definitively have functions in the logical query that are translated into something else in the Clickhouse query (think about apdex and impact). 
They are still functions there but apdex and impact do not exist as Clickhouse built in functions. The translation phase takes care of removing this function call and replace it with something clickhouse understand.
In this case they would be transformed into the arrayJoin clickhouse function.

Whether these should be functions or not, we could introduce a specific keyword node in the logical AST for these two only, but I am not sure whether it would give us any advantage over treating these two as functions.
The issue is that the join that is happening is transparent to the user, as long as that is the case (fixing which is a different project and impacts the query language and the clients) I am not sure we gain something by introducing a new type (won't take long to do that). 
I changed from Columns into FunctionCall because they were actively harming the parsing as long as we treated them as Columns, that would require introducing special cases when resolving the columns from the schema to validate the input query and have special cases during translation.

**Comment by lynnagara** [comment]
Makes sense, I did not realise we were already treating apdex and impact as FunctionCall in the logical query.

**Comment by fpacifici** [comment]
well technically we are doing this appalling thing right now. 
https://github.com/getsentry/snuba/blob/master/snuba/util.py#L78-L98

But that is not supposed to be the case and will be processed properly as a function with the ast in place

**Comment by lynnagara** [comment]
Yeah. So I guess FunctionCall *kind of* represents a Snuba function, not a ClickHouse function? Still, I guess what's weird to me is that selecting `tags_key` and `tags_value` does not look like a function in Snuba syntax either (unlike apdex and impact) even though it is one

**Comment by fpacifici** [comment]
> So I guess FunctionCall kind of represents a Snuba function, not a ClickHouse function?

Correct but only in the logical query. There a functionCall represents a snuba function. In the Physical query FunctionCall represents a clickhouse function. The fact we use the same class for both is incidental and probably temporary. In practice even if they were two different classes they would look 99% the same (I guess only the name would change).

> Still, I guess what's weird to me is that selecting tags_key and tags_value does not look like a function in Snuba syntax either
What makes it weird? (starting from the assumptions that those keywords should not exist at all but they do exist so we. have to deal with them). My reasoning here is, whether we would be treating them differently in the query processing if the user would provide them as `["tags_key", []]` instead as `tags_key`. I could not find a difference so far. I could find differences when they were represented as Column.

I don't think having a special syntax for some function call is necessarily bad (again, here the issue is that they should not have a special syntax but that's a given).
I would like the AST not to be bloated with concepts that mean the same thing even though they are represented differently in the syntax. A simpler AST makes query processing easier. (as long as it is not so simple that it conflates concepts that should be separate). 
An extension of this is having multiple syntaxes to express the same concepts, in the AST those concept could often be collapsed in the same node. See our conditions, they can be expressed as functions or with the condition syntax, but in the AST they are always functions. This allows query processors not to care about the two different syntaxes since there is no semantic difference.
I suspect Clickhouse does the same: if you write these two queries:
`SELECT  .... FROM .... WHERE a = b`
`SELECT .... FROM .... WHERE equal(a,b)`
Clickhouse will print out the same. expression in both cases in the output (the infix one), which makes me think that, despite the different syntax, they are treated internally in the same way.

So to wrap, I don't think the problem is having two syntaxes for the same ast concept (well it is a syntax issue, which means the syntax is bloated, not that the ast is representing the concepts in a wrong way) in general. 

The question is whether there is some specificity of tags_key and tags_value that should make us treating them differently than functions.

**Comment by lynnagara** [comment]
So I think the reason this is a bit of an outlier to me is in Snuba's public API a Snuba function is a different thing to the special `tags_key` and `tags_value` "columns". Snuba function is something that the user can provide using the function syntax `["apdex(arg)", null, "alias"]` or `["apdex", [arg], "alias"]` in the aggregations field. The other usages of FunctionCall line up exactly with the external API and can be used this way by the end user. `tags_key` and `tags_value` cannot be requested like this, they look more like columns in the API, except with some special behaviour that gets applied somewhere in the query processing.

I'm wondering what (if any that is) are the impacts on the external API of Snuba? Will attempting to use `tags_key` and `tags_value` using the function syntax now just start working since they get parsed into the same thing in the logical query?

**Comment by fpacifici** [comment]
> Will attempting to use tags_key and tags_value using the function syntax now just start working since they get parsed into the same thing in the logical query?

Yes they would. Which, again, I have a hard time telling whether it is a good thing or not, since they should not exist in the first place.

**Comment by fpacifici** [comment]
Actually hold on, I think I am wrong. It is not entirely true that tags_key and tags_value are not columns. "Tag" is a column in the logical schema and arguably it is a nested column there, so it would kind of make sense that `tags.key` referenced a column.
I think the question is whether it should be transformed into an arrayJoin in the logical  processing or during the translation.

Will revert the last change and actually treat this as column.

**Comment by fpacifici** [comment]
I think this way is better.
When Sentry wants to use nested columns in Snuba queries they use clickhouse array operations exposed as snuba functions (snuba does not expose anything, it blindly passes the same function to clickhouse:
https://github.com/getsentry/sentry/blob/master/src/sentry/api/serializers/snuba.py#L214

tags_key and tags_value are resolved to array functions (arrayJoin) on nested columns, thus I don't think they should be treated differently (as long as we have to support that special syntax and as a consequence we need to support arrayJoin in snuba queries).
`tags` with its structure which is composed by key and value is part of the logical schema and it should be. So I think the right way to deal with these special keywords is to translate them into array functions asap during parsing and forget that they were called tags_key and tags_value.
All optimizations can be done through query processors.

**Comment by tkaemming** [comment]
My initial thought when reading this was similar to @lynnagara's — when reading a query, there's not an obvious way to distinguish `tags_{key,value}` as a function when compared to a column, but it looks like that's changed now, so I didn't need to actually write that. 😄

I also don't think it should be the responsibility of the parser to look at this value and decide that `tags_key` is or isn't a column, even though it looks like one based on where it is provided in the syntactic structure. Then again, the parser is also is responsible for understanding the differences between column references and literals, which generally can't be syntactically distinguished either so there's not exactly precedence against it, I guess. Even so, I do think we should avoid making it more cloudy when we can. I also think this behavior seems particular to one or two specific datasets and would ideally be isolated there, rather than present in the parser where it can be applied to any/all datasets. (I'm not as deep in the details as you are here to know if there is a reason it _has_ to be in the parser, so obviously let me know if I'm missing something.)

I guess if I personally was approaching this, I'd leave them as columns at the parsing level (since they look like them, even if they aren't), and later translate them to something specialized (probably an `arrayJoin` in the logical query, through gritted teeth) later as an early logical query processing step, mostly to get this out of the parser and explicitly couple it with the dataset(s) that need it. Would that work here?

**Comment by fpacifici** [comment]
I am ok moving that logic in the first logical query processor so it is dataset specific. Now that we decided not to resolve it to a special node type nor as a special function there is really no need for it to be in the parser.
@lynnagara  ?
Will make the change as soon as we hear from Lyn.

**Comment by fpacifici** [comment]
> if there is a reason it has to be in the parser, so obviously let me know if I'm missing something.)

The only reason existed as long as this was treated as a keyword and not as a column and had to either be resolved to a special type of node for that or to a special function. But, since none of them is true, now there is no reason.

**Comment by lynnagara** [comment]
I think this approach is a lot better. Since `tags_keys`/`tags_value` is nothing but sugar for an arrayjoined `tags.key` or `tags.value` column on specific datasets having it be the same in the logical query makes sense.

**Comment by fpacifici** [comment]
now it is a query processor

File: snuba/query/parser/strings.py [hunk 1]
```diff
@@ -1,6 +1,12 @@
 import re
 
-from snuba.query.expressions import Column, Expression, Literal, SubscriptableReference
+from snuba.query.expressions import (
+    Column,
+    Expression,
+    FunctionCall,
```
Comment (tkaemming) [inline snuba/query/parser/strings.py:6]: This import is unused now.

File: snuba/query/processors/tags_expander.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
+from dataclasses import replace
+
+from snuba.query.processors import QueryProcessor
+from snuba.query.expressions import Column, Expression, FunctionCall
+from snuba.query.logical import Query
```
Comment (tkaemming) [inline snuba/query/processors/tags_expander.py:5]: nit: sort

**Review by tkaemming — APPROVED** [review]
This is a lot easier to follow now, thanks for making those changes.

**Pull request merged**